README.md
# Quantum-Safe Nginx (QSX) 

This is the official QSX tool for setting up and prototpying a quantum-secure NGINX server. 


## PREFIX 

- QSX is an extension of the **Open Quantum Safe (OQS)** project led by [Douglas Stebila](https://www.douglas.stebila.ca/research/) and [Michele Mosca](http://faculty.iqc.uwaterloo.ca/mmosca/), which aims to develop and integrate into applications quantum-safe cryptography to facilitate deployment and testing in real world contexts. In particular, OQS provides prototype integrations of liboqs into TLS and SSH, through [OpenSSL](https://github.com/open-quantum-safe/openssl) and [OpenSSH](https://github.com/open-quantum-safe/openssh-portable). 
- More information on OQS can be found [here](https://openquantumsafe.org/) and in the [associated](https://openquantumsafe.org/papers/SAC-SteMos16.pdf) [whitepapers](https://openquantumsafe.org/papers/NISTPQC-CroPaqSte19.pdf).


## Overview 

- QSX builds and configures NGINX using quantum-secure version of [OpenSSL](https://github.com/open-quantum-safe/openssl).
- The appends necessary build parameters to the NGINX configuration arguments, to the user's already existing configuration arguments. 
- NGINX is configured to use HTTPS to enable SSL, using TLS version 1.3.
- The tool provides a test script for generating a quantum-secure self-signed (or part of a chain) certificate.


## Prerequesites 
- Debian/Ubuntu machine. 
- Root user.
- OpenSSL installed version >=1.1.1
- Having an Nginx installation of >=14.1.2
- For enabling a post-quantum endpoint you should already have HTTPS enabled ([Certbot](https://github.com/certbot/certbot) is a great tool to enable HTTPS on Nginx.

### HOWTO

#### 1: Setup 

On the build machine run the following to download and build the necessary dependencies and libraries: 


```sudo bash build_setup.sh ``` or ```./build_setup.sh```

Once successfully built, the build script will ask for optional certificate generation using the `gen_cert.sh` script. 

If HTTPS is enabled then `gen_cert.sh` will edit the `nginx.conf` file using `conf_edit.py`. Finally `./build_setup.sh` calls `nginx_signal.sh` which will send the `USR2` and `HUP` signals to the Nginx master process. Documentation on Nginx processes can be found [here](http://nginx.org/en/docs/control.html)


### TESTING

If a self-signed certificate was generated and a post-quantum endpoint/port was created, then this port can be tested with a post-quantum TLS session using one of two tools. 
1. Requesting the certificate from the server using the built in OpenSSL `s_client` by running the following: 

    `apps/openssl s_client -connect host:port-curves <KEM>`
 
    List of supported algorithms that can be passed to `<KEM>` are the following default algorithms:        `kyber512:kyber768:sikep434:sikep503:frodo640aes:frodo640shake:bike1l1cpa:bike1l3cpa'
   
2. Requesting a page from server on that port using an OQS modified version of curl. Installation and usage can be found [here](https://github.com/open-quantum-safe/oqs-demos/tree/main/curl).


ChangeLog
# Summary of release changes, see commit history for more details:
# https://github.com/ibmo96/QSX/commits/master


# Release descriptions: 

* V.1.0

 - Version 1.0 of QSX functioning on Debian and Ubuntu machines. 
 - Requires Nginx installation of 1.14.2 or higher. 
 - Checks for root and nginx installaton upon start of ./build_setup.sh
 - Installs relevant dependencies based on Nginx version 1.14.2.
   - Downloads and installs zlib version 1.2.9 for the Nginx HTTP rewrite module (good for modifying functionality needed by tools like Certbot)  
 - Clones Open Quantum Safe OpenSSL and Liboqs libraries
 - Donwloads Nginx sourcefiles based on the version of the package installed Nginx. 
 - Retrieves pre-existing configure arguments provided by "nginx -V"
 - Appends --with-cc-opt, --with-ld-opt and --with-openssl arguments to the pre-existing configure arguments to implement OQS OpenSSL
 - Prompts for optional generation of a self-signed certificate (signed by a post-quantum algorithm) 
   - Can edit nginx.conf file automatically to reference self-signed certificate (if created). 
 - Sends necessary Nginx signals to update binary and configuration without server down-time. 


test.py
import sys
import re

#def test_file_stuff(file):
 #	with open(file) as f:
		#print(f.readlines())
		#list = f.readlines() 
		#print(list)
		#print(list[-1])
		#test_list = ['string one\n','string two\n']
		#list[-1:-1] = test_list
		#print(list)
	#if f.readlines()[-1].contains('/n'): 
		#print('true')

def test_re(s): 
	res = re.search('server_name(.*);', s)
	#print(res.group(1))
	res.group

def test_string_extract(s):
	start = 'server_name'
	end = ';'

	print[s.find(start)+len(start):s.rfind(end)]



def main():
	#test_string_extract('server_name    www.example.com;')
	test_re('server_name    www.example.com;')
main()


test_edit_config.py
## SOURCE --> https://github.com/open-quantum-safe/oqs-demos/blob/main/nginx/fulltest/genconfig.py

## This script performs the necessary modifications to the nginx.conf file

## Arguments needed: Port, installation directory for all libs (nginx + OQS openssl)  

import common
import os
import json
import fileinput #edit file
import shutil #copy files/folders
import re

# Script assumes nginx to have been built for this platform using build-ubuntu.sh

############# Configuration section starting here
#Default nginx.conf file location
NGX_CONF="/etc/nginx/nginx.conf"

# This is where the explanation HTML code is
TEMPLATE_FILE="index-template"

# Path where tool was cloned down and this script is running from
TOOL_PATH = os.path.dirname(os.path.realpath(__file__))

# This is where all libraries are located, default is /opt
BASEPATH="/opt"

# This is the (relative to BASEPATH) path of all certificates
PKIPATH="pki"

# This is the port where all algorithms start to be present(ed)
STARTPORT=6000

# This is the local location of the OQS-enabled OpenSSL
OPENSSL=BASEPATH+"/openssl/apps/openssl"

# This is the local OQS-OpenSSL config file
OPENSSL_CNF=BASEPATH+"/openssl/apps/openssl.cnf"

# This is the fully-qualified domain name of the server to be set up
# Ensure this is in sync with contents of ext-csr.conf file
TESTFQDN="test.openquantumsafe.org"

# This is the local folder where the root CA (key and cert) resides
CAROOTDIR="root"

# This is the file containing the SIG/KEM/port assignments
ASSIGNMENT_FILE="assignments.json"

#Template server directive file
SERVER_DIRECTIVE_FILE="/templates/server_directive.txt"


############# Functions starting here

# Generate cert chain (server and CA for a given sig alg:
# srv crt/key wind up in '<path>/<sigalg>_srv.crt|key
def gen_cert(sig_alg):
   # first check whether we already have a root CA; if not create it
   if not os.path.exists(CAROOTDIR):
           os.mkdir(CAROOTDIR)
           common.run_subprocess([OPENSSL, 'req', '-x509', '-new',
                                     '-newkey', "rsa:4096",
                                     '-keyout', os.path.join(CAROOTDIR, "CA.key"),
                                     '-out', os.path.join(CAROOTDIR, "CA.crt"),
                                     '-nodes',
                                         '-subj', '/CN=oqstest_CA',
                                         '-days', '500',
                                     '-config', OPENSSL_CNF])
           print("New root cert residing in %s." % (os.path.join(CAROOTDIR, "CA.crt")))

   # now generate suitable server keys signed by that root; adapt algorithm names to std ossl 
   if sig_alg == 'rsa3072':
       ossl_sig_alg_arg = 'rsa:3072'
   elif sig_alg == 'ecdsap256':
       common.run_subprocess([OPENSSL, "ecparam", "-name", "prime256v1", "-out", os.path.join(PKIPATH, "prime256v1.pem")])
       ossl_sig_alg_arg = 'ec:{}'.format(os.path.join(PKIPATH, "prime256v1.pem"))
   else:
       ossl_sig_alg_arg = sig_alg
   # generate server key and CSR
   common.run_subprocess([OPENSSL, 'req', '-new',
                              '-newkey', ossl_sig_alg_arg,
                              '-keyout', os.path.join(PKIPATH, '{}_srv.key'.format(sig_alg)),
                              '-out', os.path.join(PKIPATH, '{}_srv.csr'.format(sig_alg)),
                              '-nodes',
                              '-subj', '/CN='+TESTFQDN,
                              '-config', OPENSSL_CNF])
   # generate server cert off common root
   common.run_subprocess([OPENSSL, 'x509', '-req',
                                  '-in', os.path.join(PKIPATH, '{}_srv.csr'.format(sig_alg)),
                                  '-out', os.path.join(PKIPATH, '{}_srv.crt'.format(sig_alg)),
                                  '-CA', os.path.join(CAROOTDIR, 'CA.crt'),
                                  '-CAkey', os.path.join(CAROOTDIR, 'CA.key'),
                                  '-CAcreateserial',
                                  '-extfile', 'ext-csr.conf', 
                                  '-extensions', 'v3_req',
                                  '-days', '365'])




def line_search(data, strings_to_search, index):

   for i, line in enumerate(data, start=index):
      if all(x in line for x in strings_to_search):  #When we find that string, we then must search for the string where the certificate directives are:
	      if line.startswith('#'):  #If the found server directive is commented out, raise exception that HTTPS is not enabled on the server
		      continue
	      else:
		      return i, line
   return None


def append_new_server_directive(filename, data, port, server_name, cert, key, algos, html_dir, html_index):

   try:
      #create copy of server directive template file
      tmp_file =  '{}_server_directive.txt'.format(port)
      shutil.copy('/templates/server_directive.txt' , tmp_file)
      #do a search and replace for the values already encoded therin
      with open(tmp_file, 'r') as file:
         new_server_directive = file.read()
         new_server_directive = new_server_directive.replace('PORTTOBESPECIFIED', port)
         new_server_directive = new_server_directive.replace('SERVERNAMETOBE', server_name)
         new_server_directive = new_server_directive.replace('SSLCERTTOBESPECIFIED', cert)
         new_server_directive = new_server_directive.replace('SSLCERTKEYTOBESPECIFIED', key)
         new_server_directive = new_server_directive.replace('SSLCURVEALGORITHMS', algos)
         new_server_directive = new_server_directive.replace('ROOTHTMLFILELOCATION', html_dir)
         new_server_directive = new_server_directive.replace('HTMLFILENAME', html_index)

         #Insert into origin nginx.conf file (data list) right before the last '}' which ends the http directive
         data[-1:-1] = new_server_directive

      #delete tmp server directive file
      os.remove(tmp_file)
      return data
   except Exception as e:
	   print(e)


def modify_conf(filename, port, cert, key, algos,  server_name):
   try:
      with open(filename, "w") as f:
         lines = f.readlines()
         #First check if HTTPS is enabled on the server
         is_https_enabled = line_search(lines, '443 ssl',0)
         if is_https_enabled is not None: 

            port_match = ["listen", port] #must match a line in the config file that contains listen and the desired port
            port_search_res = line_search(lines, port_match, 0)
            if port_search_res is not None:

            #find ssl_certificate & ssl_certificate_key directives, input cert and key location
               cert_search_res = line_search(lines, 'ssl_certificate', port_search_res[0])
               if cert_search_res is not None:
                  #if 'ssl_certificate' is found, we input the cert and cert key locations
                  ssl_certificate = 'ssl_certificate ' + cert + ';\n'
                  lines[cert_search_res[0]] = ssl_certificate
                  ssl_certificate_key = 'ss_certificate_key ' + key + ';\n'
                  lines[cert_search_res[0]+1] = ssl_certificate_key
               else:
                  raise Exception('ssl_certificate directive not found, control that you have HTTPS enabled on your server..')
            # find ssl_ecdh_curve and input desired algos
               curves = line_search(lines, 'ssl_ecdh_curve', port_search_res[0])
               if curves is not None: 
                  ssl_ecdh_curve = 'ssl_ecdh_curve \' ' + algos + '\';\n'
                  lines[curves[0]] = ssl_ecdh_curve
                  f.writelines(lines)
                  f.close()
               else: 
                  #input ss_ecdh_curve directive just before 'location' directive
                  location = line_search(lines, 'location /', port_search_res[0])
                  lines[location[0]-1:location[0]-1] = 'ssl_ecdh_curve'
                  f.writelines(lines)
                  f.close()
            else: #If we dont find the given port in the conf file, we create a whole new server directive
            # get server_name
               if server_name is not None:
                  i, str = line_search(lines, 'server_name', 0)
                  res = re.search('server_name(.*);', str)
                  server = res.group(1).strip()
                  #generate new conf file with appended server directive
                  new_data = append_new_server_directive(filename, lines, port, server, cert, key, algos, TOOL_PATH + '/templates/html', 'index.html index.html')
               else:
                  new_data = append_new_server_directive(filename, lines, port, server_name, cert, key, algos, TOOL_PATH + '/templates/html', 'index.html index.html')
                  f.writelines(new_data)
                  f.close()
         else:
            raise Exception('Desired server directive is commented out.. Please control that HTTPS is enabled on the server...(Certbot is a good tool for that)')

   except Exception as e:
	   print(e)


def write_nginx_config(f, i, port, sig, k):
           f.write("server {\n")
           f.write("    listen              0.0.0.0:"+str(port)+" ssl;\n\n")
           f.write("    server_name         "+TESTFQDN+";\n")
           f.write("    access_log          "+BASEPATH+"logs/"+sig+"-access.log;\n")
           f.write("    error_log           "+BASEPATH+"logs/"+sig+"-error.log;\n\n")
           f.write("    ssl_certificate     "+BASEPATH+PKIPATH+"/"+sig+"_srv.crt;\n")
           f.write("    ssl_certificate_key "+BASEPATH+PKIPATH+"/"+sig+"_srv.key;\n\n")
           f.write("    ssl_protocols       TLSv1.3;\n")
           if k!="*" :  
              f.write("    ssl_ecdh_curve      "+k+";\n")
           f.write("    location / {\n")
           f.write("            ssi    on;\n")
           if k!="*" :  
              f.write("            set    $oqs_alg_name \""+sig+"-"+k+"\";\n")
           f.write("            root   html;\n")
           f.write("            index  success.html;\n")
           f.write("    }\n\n")
           # activate for more boring links-only display:
           #i.write("<li><a href=https://"+TESTFQDN+":"+str(port)+">"+sig+"/"+k+" ("+str(port)+")</a></li>\n")

           # deactivate if you don't like tables:
           i.write("<tr><td>"+sig+"</td><td>"+k+"</td><td>"+str(port)+"</td><td><a href=https://"+TESTFQDN+":"+str(port)+">"+sig+"/"+k+"</a></td></tr>\n")

           f.write("}\n\n")

# generates nginx config
def gen_conf(filename, indexbasefilename):
   port = STARTPORT
   assignments={}
   i = open(indexbasefilename, "w")
   with open(TEMPLATE_FILE, "r") as tf:
     for line in tf:
       i.write(line)

   with open(filename, "w") as f:
     # baseline config
     f.write("worker_processes  auto;\n")
     f.write("worker_rlimit_nofile  10000;\n")
     f.write("events {\n")
     f.write("    worker_connections  32000;\n")
     f.write("}\n")
     f.write("\n")
     f.write("http {\n")
     f.write("    include       conf/mime.types;\n");
     f.write("    default_type  application/octet-stream;\n")
     f.write("    keepalive_timeout  65;\n\n")
     # plain server for base information
     f.write("server {\n")
     f.write("    listen      80;\n")
     f.write("    server_name "+TESTFQDN+";\n")
     f.write("    access_log  /opt/nginx/logs/80-access.log;\n")
     f.write("    error_log   /opt/nginx/logs/80-error.log;\n\n")
     f.write("    location / {\n")
     f.write("            root   html;\n")
     f.write("            index  "+indexbasefilename+";\n")
     f.write("    }\n")
     f.write("}\n")
     f.write("server {\n")
     f.write("    listen      443 ssl;\n")
     f.write("    server_name "+TESTFQDN+";\n")
     f.write("    access_log  /opt/nginx/logs/443-access.log;\n")
     f.write("    error_log   /opt/nginx/logs/443-error.log;\n\n")
     f.write("    ssl_certificate     /etc/letsencrypt/live/"+TESTFQDN+"/fullchain.pem;\n")
     f.write("    ssl_certificate_key /etc/letsencrypt/live/"+TESTFQDN+"/privkey.pem;\n\n")
     f.write("    location / {\n")
     f.write("            root   html;\n")
     f.write("            index  "+indexbasefilename+";\n")
     f.write("    }\n")
     f.write("}\n")

     f.write("\n")
     for sig in common.signatures:
        assignments[sig]={}
        assignments[sig]["*"]=port
        write_nginx_config(f, i, port, sig, "*")
        port = port+1
        for kex in common.key_exchanges:
           # replace oqs_kem_default with X25519:
           k = "X25519" if kex=='oqs_kem_default' else kex
           write_nginx_config(f, i, port, sig, k)
           assignments[sig][k]=port
           port = port+1
     f.write("}\n")
   # deactivate if you don't like tables:
   i.write("</table>\n")

   i.write("</body></html>\n")
   i.close()
   with open(ASSIGNMENT_FILE, 'w') as outfile:
      json.dump(assignments, outfile)

def main():
   # if the argument at [1] is  a '1', it means its the first run and we are only scanning the conf file for the port (incase it exists already and we need to let the user know) 
		modify_conf('test.conf', 9003, '/ex/ex/usr/example.com/cert.pem','/ex/ex/usr/example.com/cert.key' ,'kyber512:kyber768:kyber1024', None)
	# returns 'True' or 'False' if port is found or not


  # Second run, if the argument at [1] is a '2' it means the user wants to retain their certs on that port

       # If the argument at [1] is a '3' then it means that the user wants to override the cert already defined for that port

# If the argument at [1] is a '

   #modify_conf(NGX_CONF)

main()


gen_cert.sh
#!/bin/bash

# Get domain-name
echo ""
echo "==================================================================" 
echo "What is the domain you would like to generate a quantum-safe certificate for?"
echo "=================================================================="
echo ""
read DOMAIN

## Create test certificate using OQS domain

cd $LIB_DIR && mkdir -p cacert && mkdir -p pki && set -x #prints executed commands  &&
cd $LIB_DIR/openssl/apps &&

#function gen_cert{ 
./openssl req -x509 -new -newkey $SIG_ALG -keyout $LIB_DIR/cacert/CA.key -out $LIB_DIR/cacert/CA.crt -nodes -subj "/CN=$DOMAIN" -days 365 -config openssl.cnf && 
./openssl req -new -newkey $SIG_ALG -keyout $LIB_DIR/pki/$DOMAIN\_server.key -out $LIB_DIR/pki/$DOMAIN\_server.csr -nodes -subj "/CN=$DOMAIN" -config openssl.cnf &&
./openssl x509 -req -in $LIB_DIR/pki/$DOMAIN\_server.csr -out $LIB_DIR/pki/$DOMAIN\_server.crt -CA $LIB_DIR/cacert/CA.crt -CAkey $LIB_DIR/cacert/CA.key -CAcreateserial -days 365
#}
set +x #turns off printing of executed commands

# Read which port user wants to run quantum-safe server on
echo ""
echo "=================================================================="
echo "What port would you like the quantum-safe server to listen on?"
echo " You can specify either an already used port and we will update the certificate, or an unused port."
echo "If you are unsure, press enter and port 443 will be used."
echo "=================================================================="
echo ""
PORT=443
read TMP_PORT

if [[ -z $TMP_PORT ]]
then
	echo 'No port specified, using port 443..'
else
	PORT=$TMP_PORT
        echo 'Port set to: ${PORT}'
fi


## Input generated certs and other necessary directives into nginx.conf file

ALGOS='kyber512:kyber768:sikep434:sikep503:frodo640aes:frodo640shake:bike1l1cpa:bike1l3cpa'

echo "=================================================================="
echo 'This Nginx reconfiguration supports following algorithms for key exchange: ${ALGOS}. You can visit https://github.com/open-quantum-safe/openssl for a list of all supported algorithms'
echo "=================================================================="
echo ""
	python3 $TOOL_DIR/conf_edit.py $PORT $LIB_DIR/pki/$DOMAIN\_server.crt $LIB_DIR/pki/$DOMAIN\_server.key $ALGOS $DOMAIN


conf_edit.py
## This script performs the necessary modifications to the nginx.conf file

## Arguments needed: Port, installation directory for all libs (nginx + OQS openssl)  

import common
import os
import json
import fileinput #edit file
import shutil #copy files/folders
import re
import traceback
import sys

# Script assumes nginx to have been built for this platform using build-ubuntu.sh

############# Configuration section starting here
#Default nginx.conf file location
NGX_CONF="/etc/nginx/nginx.conf"

# This is where the explanation HTML code is
TEMPLATE_FILE="index-template"

# Path where tool was cloned down and this script is running from
TOOL_PATH = os.path.dirname(os.path.realpath(__file__))

# This is where all libraries are located, default is /opt
BASEPATH="/opt"

# This is the (relative to BASEPATH) path of all certificates
PKIPATH="pki"

# This is the port where all algorithms start to be present(ed)
STARTPORT=6000

# This is the local location of the OQS-enabled OpenSSL
OPENSSL=BASEPATH+"/openssl/apps/openssl"

# This is the local OQS-OpenSSL config file
OPENSSL_CNF=BASEPATH+"/openssl/apps/openssl.cnf"

# This is the fully-qualified domain name of the server to be set up
# Ensure this is in sync with contents of ext-csr.conf file
TESTFQDN="test.openquantumsafe.org"

# This is the local folder where the root CA (key and cert) resides
CAROOTDIR="root"

# This is the file containing the SIG/KEM/port assignments
ASSIGNMENT_FILE="assignments.json"

#Template server directive file
SERVER_DIRECTIVE_FILE="/templates/server_directive.txt"


############# Functions starting here

# Generate cert chain (server and CA for a given sig alg:
# srv crt/key wind up in '<path>/<sigalg>_srv.crt|key
def gen_cert(sig_alg):
   # first check whether we already have a root CA; if not create it
   if not os.path.exists(CAROOTDIR):
           os.mkdir(CAROOTDIR)
           common.run_subprocess([OPENSSL, 'req', '-x509', '-new',
                                     '-newkey', "rsa:4096",
                                     '-keyout', os.path.join(CAROOTDIR, "CA.key"),
                                     '-out', os.path.join(CAROOTDIR, "CA.crt"),
                                     '-nodes',
                                         '-subj', '/CN=oqstest_CA',
                                         '-days', '500',
                                     '-config', OPENSSL_CNF])
           print("New root cert residing in %s." % (os.path.join(CAROOTDIR, "CA.crt")))

   # now generate suitable server keys signed by that root; adapt algorithm names to std ossl 
   if sig_alg == 'rsa3072':
       ossl_sig_alg_arg = 'rsa:3072'
   elif sig_alg == 'ecdsap256':
       common.run_subprocess([OPENSSL, "ecparam", "-name", "prime256v1", "-out", os.path.join(PKIPATH, "prime256v1.pem")])
       ossl_sig_alg_arg = 'ec:{}'.format(os.path.join(PKIPATH, "prime256v1.pem"))
   else:
       ossl_sig_alg_arg = sig_alg
   # generate server key and CSR
   common.run_subprocess([OPENSSL, 'req', '-new',
                              '-newkey', ossl_sig_alg_arg,
                              '-keyout', os.path.join(PKIPATH, '{}_srv.key'.format(sig_alg)),
                              '-out', os.path.join(PKIPATH, '{}_srv.csr'.format(sig_alg)),
                              '-nodes',
                              '-subj', '/CN='+TESTFQDN,
                              '-config', OPENSSL_CNF])
   # generate server cert off common root
   common.run_subprocess([OPENSSL, 'x509', '-req',
                                  '-in', os.path.join(PKIPATH, '{}_srv.csr'.format(sig_alg)),
                                  '-out', os.path.join(PKIPATH, '{}_srv.crt'.format(sig_alg)),
                                  '-CA', os.path.join(CAROOTDIR, 'CA.crt'),
                                  '-CAkey', os.path.join(CAROOTDIR, 'CA.key'),
                                  '-CAcreateserial',
                                  '-extfile', 'ext-csr.conf', 
                                  '-extensions', 'v3_req',
                                  '-days', '365'])



def rev_line_search(data, strings_to_search, index):
   for i, line in reversed(list(enumerate(data))):
      if all(x in line for x in strings_to_search):
	      if line.strip().startswith("#"):  #section is commented out, continue to find a section that isnt
		      continue
	      else:
		      return i, line
   return None


def line_search(data, strings_to_search, index):
   for i, line in enumerate(data, index):
      if all(x in line for x in strings_to_search):
	      if line.strip().startswith("#"):  #section is commented out, continue to find a section that isnt
		      continue
	      else:
		      return i, line
   return None


def append_new_server_directive(filename, data, port, server_name, cert, key, algos, html_dir, html_index):
   try:
      #create copy of server directive template file
      tmp_file =  '{}_server_directive.txt'.format(port)
      shutil.copy(TOOL_PATH+'/templates/server_directive.txt' , tmp_file)
      #do a search and replace for the values already encoded therin
      with open(tmp_file, 'r') as file:
         new_server_directive = file.read()
         new_server_directive = new_server_directive.replace('PORTTOBESPECIFIED', port)
         new_server_directive = new_server_directive.replace('SERVERNAMETOBE', server_name)
         new_server_directive = new_server_directive.replace('SSLCERTTOBESPECIFIED', cert)
         new_server_directive = new_server_directive.replace('SSLCERTKEYTOBESPECIFIED', key)
         new_server_directive = new_server_directive.replace('SSLCURVEALGORITHMS', algos)
         new_server_directive = new_server_directive.replace('ROOTHTMLFILELOCATION', html_dir)
         new_server_directive = new_server_directive.replace('HTMLFILENAME', html_index)

         #Insert into origin nginx.conf file (data list) right before the last '}' which ends the http directive
         res = rev_line_search(data, "}",0)
         data[res[0]:res[0]] = new_server_directive

      #delete tmp server directive file
      os.remove(tmp_file)
      return data
   except Exception as e:
	   #print(e)
      traceback.print_exc()


def modify_conf(filename, port, cert, key, algos,  server_name):
   try:
      with open(filename, "r") as f:
         lines = f.readlines()
         #First check if HTTPS is enabled on the server
         https_check = ["listen","ssl"]
         https_is_enabled = line_search(lines, https_check,0)
         if https_is_enabled is not None:

            #move on to check if desired port exists
            port_match = ["listen", port] 
            port_search_res = line_search(lines, port_match, 0)
            if port_search_res is not None:
               print('Port {} is enabled on the server, changing the certificate to a self-signed post-quantum certificate...'.format(port))
            #find ssl_certificate & ssl_certificate_key directives, input cert and key location
               cert_search_res = line_search(lines[port_search_res[0]:], 'ssl_certificate', port_search_res[0])
               if cert_search_res is not None:
                  #if 'ssl_certificate' directive is found, we input the cert and cert key locations
                  ssl_certificate = '        ssl_certificate     ' + cert + ';\n'
                  lines[cert_search_res[0]] = ssl_certificate
                  ssl_certificate_key = '        ssl_certificate_key     ' + key + ';\n'
                  lines[cert_search_res[0]+1] = ssl_certificate_key
               else:
                  raise Exception('ssl_certificate directive not found, control that you have HTTPS enabled on your server..')
            # find ssl_ecdh_curve and input desired algos
               curves = line_search(lines[port_search_res[0]:], 'ssl_ecdh_curve', port_search_res[0])
               if curves is not None: 
                  ssl_ecdh_curve = '        ssl_ecdh_curve \' ' + algos + '\';\n'
                  lines[curves[0]] = ssl_ecdh_curve
                  with open(filename, "w") as i: 
                     i.writelines(lines)
                     i.close()
               else: 
                  #input ss_ecdh_curve directive just before 'location' directive
                  location = line_search(lines[port_search_res[0]:], 'location /', port_search_res[0])
                  lines[location[0]-1:location[0]-1] = 'ssl_ecdh_curve'
                  with open(filename, "w") as i: 
                     i.writelines(lines)
                     i.close()
            else: #If we dont find the given port in the conf file, we create a whole new server directive
               print('Creating new server directive for port {}...'.format(port))
               new_data = append_new_server_directive(filename, lines, port, server_name, cert, key, algos, TOOL_PATH + '/templates', 'qsx_template.html qsx_template.htm')
               with open(filename, "w") as i: 
                  i.writelines(new_data)
                  i.close()

                  """  
                  #  --- AUTOMATION OF FINDING SERVER NAME BY SCANNING CONF FILE -------

               if server_name is None:
                  i, str = line_search(lines, 'server_name', 0)
                  res = re.search('server_name(.*);', str)
                  server = res.group(1).strip()
                  #generate new conf file with appended server directive
                  new_data = append_new_server_directive(filename, lines, port, server, cert, key, algos, TOOL_PATH + '/templates/html', 'index.html index.html')
                  with open(filename, "w") as i: 
                     i.writelines(new_data)
                     i.close()
               else:
                  new_data = append_new_server_directive(filename, lines, port, server_name, cert, key, algos, TOOL_PATH + '/templates/html', 'index.html index.html')
                  with open(filename, "w") as i: 
                     i.writelines(new_data)
                     i.close() """


         else:
            raise Exception('Please control that HTTPS is enabled on the server...(Certbot is a good tool for that)')

   except Exception as e:
	   #print(e)
      traceback.print_exc()

def write_nginx_config(f, i, port, sig, k):
           f.write("server {\n")
           f.write("    listen              0.0.0.0:"+str(port)+" ssl;\n\n")
           f.write("    server_name         "+TESTFQDN+";\n")
           f.write("    access_log          "+BASEPATH+"logs/"+sig+"-access.log;\n")
           f.write("    error_log           "+BASEPATH+"logs/"+sig+"-error.log;\n\n")
           f.write("    ssl_certificate     "+BASEPATH+PKIPATH+"/"+sig+"_srv.crt;\n")
           f.write("    ssl_certificate_key "+BASEPATH+PKIPATH+"/"+sig+"_srv.key;\n\n")
           f.write("    ssl_protocols       TLSv1.3;\n")
           if k!="*" :  
              f.write("    ssl_ecdh_curve      "+k+";\n")
           f.write("    location / {\n")
           f.write("            ssi    on;\n")
           if k!="*" :  
              f.write("            set    $oqs_alg_name \""+sig+"-"+k+"\";\n")
           f.write("            root   html;\n")
           f.write("            index  success.html;\n")
           f.write("    }\n\n")
           # activate for more boring links-only display:
           #i.write("<li><a href=https://"+TESTFQDN+":"+str(port)+">"+sig+"/"+k+" ("+str(port)+")</a></li>\n")

           # deactivate if you don't like tables:
           i.write("<tr><td>"+sig+"</td><td>"+k+"</td><td>"+str(port)+"</td><td><a href=https://"+TESTFQDN+":"+str(port)+">"+sig+"/"+k+"</a></td></tr>\n")

           f.write("}\n\n")

# generates nginx config
def gen_conf(filename, indexbasefilename):
   port = STARTPORT
   assignments={}
   i = open(indexbasefilename, "w")
   with open(TEMPLATE_FILE, "r") as tf:
     for line in tf:
       i.write(line)

   with open(filename, "w") as f:
     # baseline config
     f.write("worker_processes  auto;\n")
     f.write("worker_rlimit_nofile  10000;\n")
     f.write("events {\n")
     f.write("    worker_connections  32000;\n")
     f.write("}\n")
     f.write("\n")
     f.write("http {\n")
     f.write("    include       conf/mime.types;\n");
     f.write("    default_type  application/octet-stream;\n")
     f.write("    keepalive_timeout  65;\n\n")
     # plain server for base information
     f.write("server {\n")
     f.write("    listen      80;\n")
     f.write("    server_name "+TESTFQDN+";\n")
     f.write("    access_log  /opt/nginx/logs/80-access.log;\n")
     f.write("    error_log   /opt/nginx/logs/80-error.log;\n\n")
     f.write("    location / {\n")
     f.write("            root   html;\n")
     f.write("            index  "+indexbasefilename+";\n")
     f.write("    }\n")
     f.write("}\n")
     f.write("server {\n")
     f.write("    listen      443 ssl;\n")
     f.write("    server_name "+TESTFQDN+";\n")
     f.write("    access_log  /opt/nginx/logs/443-access.log;\n")
     f.write("    error_log   /opt/nginx/logs/443-error.log;\n\n")
     f.write("    ssl_certificate     /etc/letsencrypt/live/"+TESTFQDN+"/fullchain.pem;\n")
     f.write("    ssl_certificate_key /etc/letsencrypt/live/"+TESTFQDN+"/privkey.pem;\n\n")
     f.write("    location / {\n")
     f.write("            root   html;\n")
     f.write("            index  "+indexbasefilename+";\n")
     f.write("    }\n")
     f.write("}\n")

     f.write("\n")
     for sig in common.signatures:
        assignments[sig]={}
        assignments[sig]["*"]=port
        write_nginx_config(f, i, port, sig, "*")
        port = port+1
        for kex in common.key_exchanges:
           # replace oqs_kem_default with X25519:
           k = "X25519" if kex=='oqs_kem_default' else kex
           write_nginx_config(f, i, port, sig, k)
           assignments[sig][k]=port
           port = port+1
     f.write("}\n")
   # deactivate if you don't like tables:
   i.write("</table>\n")

   i.write("</body></html>\n")
   i.close()
   with open(ASSIGNMENT_FILE, 'w') as outfile:
      json.dump(assignments, outfile)

def main():
   #modify_conf(TOOL_PATH + '/test.conf','9003', 'v2/example.com/cert.pem','v2/example.com/cert.key' ,'kyber512:kyber768:kyber1024', 'www.quantumsafenginx.com quantumsafenginx.com')
   modify_conf('/etc/nginx/nginx.conf', sys.argv[1], sys.argv[2], sys.argv[3], sys.argv[4], sys.argv[5])
  # If the argument at [1] is a '1' it means the user wants to retain their certs on that port
#append_new_server_directive('/etc/nginx/nginx.conf', None, , server_name, cert, key, algos, html_dir, html_index)
  # If the argument at [1] is a '2' then it means that the user wants to override the cert already defined for that port

main()
